libs2eplugins/src/s2e/Plugins/Models/BaseFunctionModels.cpp

```diff
@@ -692,42 +692,67 @@ bool BaseFunctionModels::strcatHelper(S2EExecutionState *state, const uint64_t s
 
     return true;
 }
-
 bool BaseFunctionModels::strstrHelper(S2EExecutionState *state, uint64_t haystackAddr, uint64_t needleAddr, ref<Expr> &retExpr, uint32_t byte_width) {
-    getWarningsStream(state) << "Enter strstr yes\n";
+    getWarningsStream(state) << "Entering strstrHelper\n";
+
     size_t haystackLen, needleLen;
-    if (!findNullCharWithWidth(state, haystackAddr, haystackLen, byte_width) || !findNullCharWithWidth(state, needleAddr, needleLen, byte_width)) {
+    if (!findNullCharWithWidth(state, haystackAddr, haystackLen, byte_width) ||
+        !findNullCharWithWidth(state, needleAddr, needleLen, byte_width)) {
         getWarningsStream(state) << "Failed to find nullptr in haystack or needle\n";
         return false;
     }
 
-    const Expr::Width ret_width = state->getPointerSize() * CHAR_BIT;
+    const Expr::Width pointerWidth = state->getPointerSize() * CHAR_BIT;
+
     if (needleLen == 0) {
-        retExpr = E_CONST(haystackAddr, ret_width);
-        return true;
+        retExpr = E_CONST(haystackAddr, pointerWidth);
+        return true;  
     }
+    const ref<Expr> nullExpr = E_CONST(0, pointerWidth);
 
-    const ref<Expr> nullExpr = E_CONST(0, ret_width);
-
-    ref<Expr> finalExpr = nullExpr; // Start with the assumption that we will not find the needle
-    for (size_t i = 0; i <= haystackLen - needleLen; i += byte_width) {
-        uint64_t strAddrs[2] = {haystackAddr + i, needleAddr};
-        ref<Expr> memcmpResult;
-
-        // Use strncmpHelper to compare the substring of haystack with needle
-        if (!memcmpHelper(state, strAddrs, needleLen, memcmpResult)) {
-            getWarningsStream(state) << "Failed to compare substring of haystack with needle " << needleLen << "\n";
+    std::vector<size_t> badCharSkip(256, needleLen);  // Bad character skip array. initialized to needleLen for all entry
+    uint8_t charByte = 0;  // Temporary variable for reading bytes
+    for (size_t i = 0; i < needleLen - 1; ++i) {
+        uint64_t readAddr = needleAddr + i * byte_width;
+        if (!state->mem()->read(readAddr, &charByte, VirtualAddress, true)) {
+            getWarningsStream(state) << "Failed to read byte at address " << hexval(readAddr) << "\n";
             return false;
         }
+        badCharSkip[charByte] = needleLen - 1 - i; // set each of the needle char to corresponding skip len
```

Comment:
What if the `charByte` is symbolic? What will `s2e` do? And why don't we consider the character's with in this case? (why do we just treat every char as one byte)?

Reply:
it won't be symbolic, as mem()->read() will silently concretize the value. And yes, that's a good point. We should consider character width here

Reply:
fixed